File: abm/onLatticeCA_jar/OnLatticeModel.java
package onLatticeCA_jar;

import HAL.lib.CommandLine;
import java.io.File;
import static HAL.Util.LoadState;

// ------------------------------------------------------------------------------------------------------------
// Main "Wrapper" class that manages input/output and the simulation.
// ------------------------------------------------------------------------------------------------------------
@CommandLine.Command(name = "Game Assay Simulator",
        mixinStandardHelpOptions = true,
        showDefaultValues = true,
        description = "A 2D on-lattice agent-based model that simulates 2 cell populations growing together in a domain under a given drug schedule.")
public class OnLatticeModel implements Runnable {
    /* ========================================================================
     * --- Parameters ----
     * ======================================================================== */
    // Parse variables from command line using piclocli (http://picocli.info)
    OnLatticeGrid model = new OnLatticeGrid(); // Temporary placeholder so can load defaults from the base class
    // ------------------------- Config -------------------------
    @CommandLine.Option(names = { "-h", "--help" }, usageHelp = true, description = "Display a help message and exit.") 
    boolean helpRequested;
    @CommandLine.Option(names = { "-s", "--seed"}, description="Random number seed.") 
    int seed = model.seed;
    @CommandLine.Option(names = { "--headless" }, description = "Whether or not to run in headless mode (used for running on cluster).") 
    boolean headless = true;

    // ------------------------- Experimental Setup -------------------------
    @CommandLine.Option(names = { "-xDim", "--xDim"}, description="x-dimension of domain (in lattice sites)") 
    int xDim = model.xDim;
    @CommandLine.Option(names = { "-yDim", "--yDim"}, description="y-dimension of domain (in lattice sites)") 
    int yDim = model.yDim;
    @CommandLine.Option(names = { "-n", "--nReplicates"}, description="Number of replicates.") 
    int nReplicates = model.nReplicates;
    @CommandLine.Option(names = { "-N0", "--initialSize"}, description="Total initial cell number (absolute)") 
    double initialSize = model.initialSize;
    @CommandLine.Option(names = { "-n0", "--initialSizeProp"}, description="Initial cell density relative to (physical) carrying capacity") 
    double initialSizeProp = model.initialSizeProp;
    @CommandLine.Option(names = { "-fR", "--rFrac"}, description="Initial resistance fraction in [0,1]") 
    double rFrac = model.rFrac;    
    @CommandLine.Option(names = { "-tEnd", "--tEnd"}, description="End time in days.") 
    double tEnd = model.tEnd;
    @CommandLine.Option(names = { "-dt", "--dt"}, description="Time step in days.") 
    double dt = model.dt;
    @CommandLine.Option(names = { "-treatmentScheduleList", "--treatmentScheduleList"}, description="Treatment schedule in format {{tStart, tEnd, drugConcentration}}.") 
    String treatmentScheduleList_string;

    // ------------------------- Cell Properties -------------------------
    @CommandLine.Option(names = { "-r_S", "--divisionRate_S"}, description="Proliferation rate of sensitive cells in d^-1. Proliferation will be attempted at this rate.") 
    double divisionRate_S = model.divisionRate_S;
    @CommandLine.Option(names = { "-r_R", "--divisionRate_R"}, description="Proliferation rate of sensitive cells in d^-1. Proliferation will be attempted at this rate.") 
    double divisionRate_R = model.divisionRate_R;
    @CommandLine.Option(names = { "-m_S", "--movementRate_S"}, description="Movement rate of sensitive cells in d^-1. Movement will be attempted at this rate.") 
    double movementRate_S = model.movementRate_S;
    @CommandLine.Option(names = { "-m_R", "--movementRate_R"}, description="Movement rate of resistant cells in d^-1. Movement will be attempted at this rate.") 
    double movementRate_R = model.movementRate_R;
    @CommandLine.Option(names = { "-d_S", "--deathRate_S"}, description="Natural death rate of sensitive cells in d^-1. Mean life span of a cell will be 1/k.") 
    double deathRate_S = model.deathRate_S;
    @CommandLine.Option(names = { "-d_R", "--deathRate_R"}, description="Natural death rate of resistant cells in d^-1. Mean life span of a cell will be 1/k.") 
    double deathRate_R = model.deathRate_R;
    @CommandLine.Option(names = { "-d", "--drugKillProportion"}, description="Drug induced death rate in d^-1.") 
    double drugKillProportion = model.drugKillProportion;

    // ------------------------- Output - Text -------------------------
    @CommandLine.Option(names = { "-v", "--verboseLevel"}, description="Verbosity (0-2).") 
    int verboseLevel = model.verboseLevel;
    @CommandLine.Option(names = { "-printfreq", "--printFrequency"}, description="Frequency at which output is printed to the screen.") 
    double printFrequency = model.printFrequency;
    @CommandLine.Option(names = { "-logf", "--logCellCountFrequency"}, description="Frequency (in time units) at which the cell counts are written to file. <0 indicates no logging.") 
    double logCellCountFrequency = model.logCellCountFrequency;
    // @CommandLine.Option(names = { "--extraSimulationInfo"}, description="Additional columns to be printed into the log file. Useful when want to annotate experiment to facilitate later analysis.") 
    // double[] extraSimulationInfo = model.extraSimulationInfo;
    // @CommandLine.Option(names = { "--extraSimulationInfoNames"}, description="Column headers to be used for additional columns to be printed into the log file.") 
    // String[] extraSimulationInfoNames = model.extraSimulationInfoNames;
    @CommandLine.Option(names = { "--outDir"}, description="Directory which to save output files to.") 
    String outDir = "./tmp/";
    
    // ------------------------- Output - Visualisation -------------------------
    @CommandLine.Option(names = { "-visualiseB", "--visualiseB"}, description="Whether or not to show visualization.")
    Boolean visualiseB = model.visualiseB;
    @CommandLine.Option(names = { "--scaleFactor"}, description="Scale factor used to display grid.") 
    int scaleFactor = model.scaleFactor;
    @CommandLine.Option(names = { "--pause"}, description="Pause betwween time steps to slow simulation.") 
    int pause = model.pause;
    @CommandLine.Option(names = { "--imageFrequency"}, description="Frequency at which an image of the tumour is saved. Negative number turns it off.") 
    int imageFrequency = model.imageFrequency;
    @CommandLine.Option(names = { "--imageOutDir"}, description="Directory which to save images to.") 
    String imageOutDir = model.imageOutDir;

    // ------------------------- Output - Model File -------------------------
    @CommandLine.Option(names = { "--saveModelState"}, description="Whether or not to save the model object at the end of the simulation.") 
    Boolean saveModelState = model.saveModelState;
    // @CommandLine.Option(names = { "--fromScratch"}, description="Whether or not to run simulation from scratch or from a model file.") 
    // Boolean fromScratch = model.fromScratch; // #TODO: Can I clean this up?
    @CommandLine.Option(names = { "--savedModelFileName"}, description="Name of model file to load when continuing a previous run.") 
    String savedModelFileName = model.savedModelFileName;

    // ------------------------------------------------------------------------------------------------------------
    /*
     * Main runner function
     */
    public void run(){
        // ------------------------- 1. Setup -------------------------
        assert !helpRequested; // Print help message if requested
        if (headless) {System.setProperty("java.awt.headless", "true");} // Make headless if necessary
        OnLatticeGrid myModel;
        double[] paramArr;
        String currSavedModelFileName;
        String outFName;
        Boolean fromScratch = savedModelFileName==null;
        new File(outDir).mkdirs(); // Set up environment

        // Parse treatment schedule. If none set, simulate growth until the end of the simulation.
        double[][] treatmentScheduleList;
        if (treatmentScheduleList_string == null) {
            treatmentScheduleList = new double[][] {{0, tEnd, 0}};
        } else {
            // Parse command line input
            String[] tmpList = treatmentScheduleList_string.split("\\[");
            String[] processed_justNumbers;
            treatmentScheduleList = new double[tmpList.length-2][3];
            for (int k=2; k<tmpList.length; k++) {
                processed_justNumbers = tmpList[k].split("\\]");
                processed_justNumbers = processed_justNumbers[0].split(",");
                for (int kk=0; kk<processed_justNumbers.length; kk++) {
                    treatmentScheduleList[k-2][kk] = Double.parseDouble(processed_justNumbers[kk]);
                }
            }
        }

        // Parse initial size information. If fraction provided use that. Otherwise use absolute numbers
        if (initialSizeProp > 0) {
            initialSize = (int) (initialSizeProp * xDim * yDim);
        }
        int[] initPopSizeArr = new int[]{(int) Math.floor(initialSize * (1 - rFrac)), (int) Math.ceil(initialSize * rFrac)};

        // Prepare storage and indexing for each replicate
        int[] replicateIdList = new int[] {seed};
        int replicateId;
        if (nReplicates>1) {
            replicateIdList = new int[nReplicates];
            for (int i=0; i<nReplicates; i++) {replicateIdList[i]=i;}
        }

        // ------------------------------------------------------------------------
        // Main simulation loop.
        // ------------------------------------------------------------------------
        for (int replicateIdx = 0; replicateIdx < nReplicates; replicateIdx++) {
            replicateId = replicateIdList[replicateIdx];
            currSavedModelFileName = savedModelFileName==null? outDir + "RepId_" + replicateId + ".bin": savedModelFileName;

            // Models can be created either from scratch, or be loaded from a previous run
            if (fromScratch) {
                // Set up the simulation
                paramArr = new double[]{divisionRate_S, divisionRate_R, movementRate_S, movementRate_R,
                        deathRate_S, deathRate_R, drugKillProportion};
                myModel = new OnLatticeGrid(xDim, yDim, paramArr, dt);
                myModel.tEnd = tEnd;
                myModel.visualiseB = visualiseB;
                myModel.verboseLevel = verboseLevel;
                myModel.pause = pause;

                // Set the random number seed. Behaviour depends no whether this is a single run or part of a series of nReplicate runs. By default will assign every replicate the value ```seed=replicateId```
                if (seed != -1) {
                    if (nReplicates == 1) {
                        myModel.SetSeed(seed);
                    } else {
                        myModel.SetSeed(replicateId, seed);
                    }
                } else {
                    myModel.SetSeed(replicateId);
                }

                // Set the logging behaviour
                myModel.SetExtraSimulationInfo(new String[]{"ReplicateId", "InitSize", "RFrac"},
                        new double[]{replicateId, initialSizeProp, rFrac});
                outFName = outDir + "RepId_" + replicateId + ".csv";
                if (imageFrequency > 0) {
                    myModel.visualiseB = true;
                    String currImgOutDir = imageOutDir + "RepId_" + replicateId;
                    new File(currImgOutDir).mkdirs();
                    myModel.ConfigureImaging(currImgOutDir, imageFrequency);
                }

                // Initialise the simulation
                myModel.InitialiseCellLog(outFName, dt);
                myModel.SetInitialState(initPopSizeArr);
            } else {
                myModel = LoadState(currSavedModelFileName);
            }

            // Run the simulation
            myModel.SetTreatmentSchedule(treatmentScheduleList);
            myModel.Run();
            myModel.Close();
            if (saveModelState) {myModel.SaveModelState(currSavedModelFileName);}
        }
    }

    public static void main(String[] args) {
        int exitCode = new CommandLine(new OnLatticeModel()).execute(args); 
        System.exit(exitCode);
    }
}

File: utils/OnLatticeModel.py
# ====================================================================================
# Class to simulate spheroid growth using one of 6 ODE models
# ====================================================================================
import numpy as np
import pandas as pd
import math
import os
import sys
if 'matplotlib' not in sys.modules:
    import matplotlib as mpl
    mpl.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(style="white")
# ====================================================================================
class OnLatticeModel():
    def __init__(self, **kwargs):
        # Initialise parameters
        self.modelConfigDic = {
                               # -------------------- Experimental Setup --------------------
                               "xDim": 100, "yDim": 100, # Dimensions of the domain (in lattice sites)
                               "nReplicates": 10, # Number of replicates
                               "seed": -1, # Random number seed
                               "initialSize": 250, # Total initial cell number (absolute)
                               "initialSizeProp": 0, # Initial cell density relative to (physical) carrying capacity; takes precedence over the absolute number if it is specified.
                               "rFrac": 0.05, # Initial resistance fraction in [0,1]
                               "tEnd": 5, # End time in days
                               "dt": 1, # Time step in days
                               # TODO: Test whether the schedule gets passed correctly
                               "treatmentScheduleList": None, # Treatment schedule in format [[tStart, tEnd, drugConcentration]]
                               # -------------------- Cell Properties --------------------
                               "divisionRate_S": 0.027, "divisionRate_R": 0.027, # Proliferation rate of cells in d^-1.
                               "movementRate_S": 0.0, "movementRate_R": 0.0, # Movement rate of cells in d^-1.
                               "deathRate_S": 0.027, "deathRate_R": 0.027, # Natural death rate of cells in d^-1.
                               "drugKillProportion": 0.75, # Probability that dividing sensitive cell is killed by drug.
                               "turnover": None, "cost": None, # Shorthand way of specifying cost and turnover
                               # -------------------- Output - Text --------------------
                               "verboseLevel": 0, "printFrequency": 100, # Command line verbosity. This goes to stdout.
                               "logCellCountFrequency": 1, # Frequency (in time units) at which the cell counts are written to file. <0 indicates no logging.
                               "outDir": "./tmp/", # Directory which to save output files to.
                               # -------------------- Output - Visualisation --------------------
                               # TODO: Clean up how visualisation are managed (and headless vs visualiseB stuff)
                            #    "headless": True, # Whether to show graphics window. Want this off to run on cluster etc
                            #    "visualiseB": False, # Whether to generate visualisation. Turn off for speed increase.
                               "scaleFactor": 2, "pause": 0, # Visualsation configs
                               "imageOutDir": "./tmp/", "imageFrequency": -1, # Saving of simulation images. Negative freq turns it off.") 
                               # -------------------- Output - Model File --------------------
                               # TODO: Get this implemented
                               "saveModelState": False, # Whether or not to save the model object at the end of the simulation.
                               "savedModelFileName": None # Name of model file to load when continuing a previous run.
                               }
        self.jarFileName = kwargs.get('jarFileName', './onLatticeModel.jar')
        self.resultsDf = None

        # Set the parameters
        self.SetParams(**kwargs)

        # Configure the solver
        self.suppressOutputB = kwargs.get('suppressOutputB',
                                          False)  # If true, suppress output of ODE solver (including warning messages)

    # =========================================================================================
    # Function to set the parameters
    def SetParams(self, **kwargs):
        for key in self.modelConfigDic.keys():
            self.modelConfigDic[key] = kwargs.get(key, self.modelConfigDic[key])
        # The 'cost' and 'turnover' parameters provide shorthands to set the global deathrate and resistant cell proliferation rate
        if self.modelConfigDic['turnover'] != None:
            self.modelConfigDic['deathRate_S'] = self.modelConfigDic['turnover']*self.modelConfigDic['divisionRate_S']
            self.modelConfigDic['deathRate_R'] = self.modelConfigDic['turnover']*self.modelConfigDic['divisionRate_S']
        if self.modelConfigDic['cost'] != None:
            self.modelConfigDic['divisionRate_R'] = (1-self.modelConfigDic['cost'])*self.modelConfigDic['divisionRate_S']
        # Ensure that the paths for the output directories have a file separator at the end so that
        # java then creates the correct file names.
        self.modelConfigDic['outDir'] = os.path.join(self.modelConfigDic['outDir'], "")
        if 'imageOutDir' in self.modelConfigDic: self.modelConfigDic['imageOutDir'] = os.path.join(self.modelConfigDic['imageOutDir'], "")

    def ConvertTreatmentScheduleToStr(self, treatmentScheduleList):
        treatmentScheduleStr = "["
        for interval in treatmentScheduleList:
            treatmentScheduleStr += "[%1.2f,%1.2f,%1.2f]," % tuple(interval)
        treatmentScheduleStr = treatmentScheduleStr[:-1]
        treatmentScheduleStr += "]"
        return treatmentScheduleStr

    def RunSimulation(self, printCommand=False):
        '''
        Parses arguments into command line command and run this command in the terminal
        '''
        argStr = " "
        for var in self.modelConfigDic.keys():
            if self.modelConfigDic[var] is None: continue # For vars set to None use default vals from java side
            if var in ["cost", "turnover"]: continue # Certain parameters are only used in the wrapper and not passed to java
            if isinstance(self.modelConfigDic[var], bool): 
                if self.modelConfigDic[var]: argStr += "--%s " % (var)
            else:
                argStr += "--%s %s " % (var, self.modelConfigDic[var])
        if printCommand: print("java -jar %s" % self.jarFileName + argStr)
        os.system("java -jar %s" % self.jarFileName + argStr)

    def LoadSimulations(self, normalise=True):
        tmpList = []
        replicateIdList = range(self.modelConfigDic['nReplicates']) if self.modelConfigDic['nReplicates']>1 else [self.modelConfigDic['seed']]
        for replicateId in replicateIdList:
            currDfName = os.path.join(self.modelConfigDic['outDir'], "RepId_%d.csv" % (replicateId))
            tmpDf = pd.read_csv(currDfName)
            tmpDf = tmpDf[['Time', 'NCells_S', 'NCells_R', 'NCells', 'DrugConcentration']]
            tmpDf['ReplicateId'] = replicateId
            tmpList.append(tmpDf)
        resultsDf = pd.concat(tmpList)
        resultsDf.rename(columns={"NCells": "TumourSize",
                                  "NCells_S": "S",
                                  "NCells_R": "R"}, inplace=True)
        if normalise:
            resultsDf['TumourSize'] /= 1e4
            resultsDf['S'] /= 1e4
            resultsDf['R'] /= 1e4
        return resultsDf

    def NormaliseToInitialSize(self, dataDf):
        dataDf['S'] /= dataDf.TumourSize.iloc[0]
        dataDf['R'] /= dataDf.TumourSize.iloc[0]
        dataDf['TumourSize'] /= dataDf.TumourSize.iloc[0]

    # =========================================================================================
    # Function to simulate the model
    def Simulate(self, treatmentScheduleList=None, scaleTumourVolume=False, printCommand=False, **kwargs):
        # Allow configuring the solver at this point as well
        self.jarFileName = kwargs.get('jarFileName', self.jarFileName)
        if treatmentScheduleList is not None: self.modelConfigDic["treatmentScheduleList"] = self.ConvertTreatmentScheduleToStr(treatmentScheduleList)
        # if (self.modelConfigDic['imageFreq']==-1) and ('imageOutDir' in self.modelConfigDic): self.modelConfigDic.pop('imageOutDir')
        self.SetParams() # Calling this to make sure parameters and file paths are properly checked in case they were reset at any point

        # Run the simulations
        self.RunSimulation(printCommand=printCommand)
        # self.SetParams(fromScratch=False) # TODO: get continuation to work again

        # Load data
        self.resultsDf = self.LoadSimulations()
        self.resultsDf = self.resultsDf.groupby(by="Time").mean()
        self.resultsDf.reset_index(inplace=True)
        self.resultsDf.drop(columns="ReplicateId", inplace=True)
        if scaleTumourVolume: self.NormaliseToInitialSize(self.resultsDf)

    # =========================================================================================
    # Function to plot the model predictions
    def Plot(self, scaleTumourVolume=True, aggregateData=True, progressBar=False, drugBarPosition=0.85,
             xlim=None, ylim=1.3, y2lim=1, decorateX=True, decorateY=True, axisLabels=False, markersize=10,
             labelsize=28,
             titleStr="", ax=None, figsize=(10, 8), outName=None, **kwargs):
        if ax is None:
            fig = plt.figure(figsize=figsize)
            ax = fig.add_subplot(111)

        # Load data
        predictionDf = self.LoadSimulations()
        if scaleTumourVolume: self.NormaliseToInitialSize(predictionDf)

        # Plot the size the we will see on the images
        sns.lineplot(x="Time", y="TumourSize", ci='sd',
                     lw=kwargs.get('linewidthA', kwargs.get('linewidth', 7)),
                     color=kwargs.get('colorA', '#094486'),
                     estimator='mean' if aggregateData else None, legend=False,
                     data=predictionDf, ax=ax)

        # Plot the individual populations
        sns.lineplot(x="Time", y="S", ci='sd',
                     lw=kwargs.get('linewidth', 7), color=kwargs.get('colorS', "#0F4C13"),
                     estimator='mean' if aggregateData else None, legend=False,
                     data=predictionDf, ax=ax)
        ax.lines[1].set_linestyle(kwargs.get('linestyleS', '--'))
        sns.lineplot(x="Time", y="R", ci='sd',
                     lw=kwargs.get('linewidth', 7), color=kwargs.get('colorR', '#710303'),
                     estimator='mean' if aggregateData else None, legend=False,
                     data=predictionDf, ax=ax)
        ax.lines[2].set_linestyle(kwargs.get('linestyleR', '-.'))

        # Plot the drug concentration
        ax2 = ax.twinx()  # instantiate a second axes that shares the same x-axis
        exampleReplicateId = predictionDf.ReplicateId.unique()[0]
        timeVec = predictionDf.Time[predictionDf.ReplicateId == exampleReplicateId]
        drugConcentrationVec = predictionDf.DrugConcentration[predictionDf.ReplicateId == exampleReplicateId]
        drugConcentrationVec = drugConcentrationVec / (1 - drugBarPosition) + drugBarPosition
        ax2.fill_between(timeVec,
                         drugBarPosition, drugConcentrationVec, color="black",
                         alpha=1., label="Drug Concentration")
        ax2.axis("off")

        # Format the plot
        if xlim is not None: ax.set_xlim(0, xlim)
        ax.set_ylim(0, ylim)
        ax2.set_ylim([0, y2lim])
        ax.set_xlabel("Time in Days" if axisLabels else "", fontdict={'fontsize': 28})
        ax.set_ylabel("PSA (Normalised)" if axisLabels else "", fontdict={'fontsize': 28})
        ax.set_title(titleStr)
        ax.tick_params(labelsize=labelsize)
        ax2.tick_params(labelsize=labelsize)
        if not decorateX:
            ax.set_xticklabels("")
        if not decorateY:
            ax.set_yticklabels("")
        plt.tight_layout()
        if outName is not None: plt.savefig(outName)

    # =========================================================================================
    # # Function to plot the model predictions
    # def Make_Gif(self, treatmentScheduleList=None, seed=1, imageFreq=10, printCommand=False,
    #          titleStr="", ax=None, figsize=(10, 8), outName=None, **kwargs):
    #     if ax is None:
    #         fig = plt.figure(figsize=figsize)
    #         ax = fig.add_subplot(111)
    #
    #     # Create images for Gif
    #     configDic_gifMaker = self.modelConfigDic.copy()
    #     if treatmentScheduleList is None:
    #         if self.modelConfigDic["treatmentScheduleList"] is None:
    #             raise ValueError("No treatment schedule defined!")
    #     else:
    #         configDic_gifMaker["treatmentScheduleList"] = self.ConvertTreatmentScheduleToStr(treatmentScheduleList)
    #     configDic_gifMaker['seed'] = seed
    #     configDic_gifMaker['nReplicates'] = 1
    #     configDic_gifMaker['imageFreq'] = imageFreq
    #     RunSimulation(configDic_gifMaker, jarFileName=self.jarFileName, printCommand=printCommand)
    #
    #     # Format the plot
    #     ax.set_title(titleStr)
    #     plt.tight_layout()
    #     if outName is not None: plt.savefig(outName)